File: examples/spanner-client/php/cli_options.php
<?php

declare(strict_types=1);

final class CliOptions
{
    public function __construct(
        public readonly string $queryMode,
        public readonly string $project,
        public readonly string $instance,
        public readonly string $database,
        public readonly string $sql,
    ) {
    }

    public static function parse(array $argv = []): self
    {
        $args = $argv !== [] ? $argv : array_slice($_SERVER['argv'] ?? [], 1);
        $queryMode = strtoupper(self::envOrDefault('SPANNER_QUERY_MODE', 'PLAN'));
        $project = self::envOrNull('SPANNER_PROJECT_ID');
        $instance = self::envOrNull('SPANNER_INSTANCE_ID');
        $database = self::envOrNull('SPANNER_DATABASE_ID');
        $query = self::envOrNull('SPANNER_QUERY');
        $queryFile = self::envOrNull('SPANNER_QUERY_FILE');

        for ($i = 0; $i < count($args); $i++) {
            $arg = $args[$i];
            if ($arg === '-h' || $arg === '--help') {
                self::printUsage();
                exit(0);
            }
            if ($arg === '--query-mode' && isset($args[$i + 1])) {
                $queryMode = strtoupper($args[++$i]);
                continue;
            }
            if ($arg === '--project' && isset($args[$i + 1])) {
                $project = $args[++$i];
                continue;
            }
            if ($arg === '--instance' && isset($args[$i + 1])) {
                $instance = $args[++$i];
                continue;
            }
            if ($arg === '--database' && isset($args[$i + 1])) {
                $database = $args[++$i];
                continue;
            }
            if ($arg === '--query' && isset($args[$i + 1])) {
                $query = $args[++$i];
                continue;
            }
            if ($arg === '--query-file' && isset($args[$i + 1])) {
                $queryFile = $args[++$i];
                continue;
            }
            throw new InvalidArgumentException("unknown argument: {$arg}");
        }

        if ($queryMode !== 'PLAN' && $queryMode !== 'PROFILE') {
            throw new InvalidArgumentException("query mode must be PLAN or PROFILE, got: {$queryMode}");
        }
        self::requireValue('project', 'SPANNER_PROJECT_ID', $project);
        self::requireValue('instance', 'SPANNER_INSTANCE_ID', $instance);
        self::requireValue('database', 'SPANNER_DATABASE_ID', $database);

        return new self(
            $queryMode,
            $project,
            $instance,
            $database,
            self::loadSql($query, $queryFile),
        );
    }

    private static function requireValue(string $flag, string $env, ?string $value): void
    {
        if ($value === null || trim($value) === '') {
            throw new InvalidArgumentException("missing required value: set --{$flag} or {$env}");
        }
    }

    private static function loadSql(?string $query, ?string $queryFile): string
    {
        if ($query !== null && trim($query) !== '') {
            return trim($query);
        }
        $path = ($queryFile !== null && trim($queryFile) !== '') ? $queryFile : __DIR__ . '/../query.sql';
        $sql = file_get_contents($path);
        if ($sql === false) {
            throw new RuntimeException("failed to read query file: {$path}");
        }

        return trim($sql);
    }

    private static function envOrNull(string $name): ?string
    {
        $value = getenv($name);
        if (!is_string($value) || trim($value) === '') {
            return null;
        }

        return trim($value);
    }

    private static function envOrDefault(string $name, string $default): string
    {
        return self::envOrNull($name) ?? $default;
    }

    private static function printUsage(): void
    {
        fwrite(STDERR, <<<'USAGE'
usage: analyze_and_render.php [options]
  --query-mode PLAN|PROFILE   Spanner execute-sql mode (default: PLAN)
  --project PROJECT           GCP project id
  --instance INSTANCE         Spanner instance id
  --database DATABASE         Spanner database id
  --query SQL                 SQL text (overrides --query-file)
  --query-file PATH           SQL file (default: ../query.sql)

Environment (when flags omitted):
  SPANNER_QUERY_MODE, SPANNER_PROJECT_ID, SPANNER_INSTANCE_ID,
  SPANNER_DATABASE_ID, SPANNER_QUERY, SPANNER_QUERY_FILE

USAGE);
    }
}

File: examples/spanner-client/java/src/main/java/io/spannerplan/examples/CliOptions.java
package io.spannerplan.examples;

import java.nio.file.Files;
import java.nio.file.Path;

/** CLI flags and environment for Spanner client examples. */
public final class CliOptions {
  public final String queryMode;
  public final String project;
  public final String instance;
  public final String database;
  public final String sql;

  private CliOptions(String queryMode, String project, String instance, String database, String sql) {
    this.queryMode = queryMode;
    this.project = project;
    this.instance = instance;
    this.database = database;
    this.sql = sql;
  }

  public static CliOptions parse(String[] args) throws Exception {
    String queryMode = envOrDefault("SPANNER_QUERY_MODE", "PLAN").toUpperCase();
    String project = envOrNull("SPANNER_PROJECT_ID");
    String instance = envOrNull("SPANNER_INSTANCE_ID");
    String database = envOrNull("SPANNER_DATABASE_ID");
    String query = envOrNull("SPANNER_QUERY");
    String queryFile = envOrNull("SPANNER_QUERY_FILE");

    for (int i = 0; i < args.length; i++) {
      String arg = args[i];
      if ("-h".equals(arg) || "--help".equals(arg)) {
        printUsage();
        System.exit(0);
      }
      if ("--query-mode".equals(arg) && i + 1 < args.length) {
        queryMode = args[++i].toUpperCase();
        continue;
      }
      if ("--project".equals(arg) && i + 1 < args.length) {
        project = args[++i];
        continue;
      }
      if ("--instance".equals(arg) && i + 1 < args.length) {
        instance = args[++i];
        continue;
      }
      if ("--database".equals(arg) && i + 1 < args.length) {
        database = args[++i];
        continue;
      }
      if ("--query".equals(arg) && i + 1 < args.length) {
        query = args[++i];
        continue;
      }
      if ("--query-file".equals(arg) && i + 1 < args.length) {
        queryFile = args[++i];
        continue;
      }
      throw new IllegalArgumentException("unknown argument: " + arg);
    }

    if (!"PLAN".equals(queryMode) && !"PROFILE".equals(queryMode)) {
      throw new IllegalArgumentException("query mode must be PLAN or PROFILE, got: " + queryMode);
    }
    requireValue("project", "SPANNER_PROJECT_ID", project);
    requireValue("instance", "SPANNER_INSTANCE_ID", instance);
    requireValue("database", "SPANNER_DATABASE_ID", database);

    String sql = loadSql(query, queryFile);
    return new CliOptions(queryMode, project, instance, database, sql);
  }

  private static void requireValue(String flag, String env, String value) {
    if (value == null || value.isBlank()) {
      throw new IllegalStateException(
          "missing required value: set --" + flag + " or " + env);
    }
  }

  private static String loadSql(String query, String queryFile) throws Exception {
    if (query != null && !query.isBlank()) {
      return query.trim();
    }
    Path path =
        queryFile != null && !queryFile.isBlank()
            ? Path.of(queryFile)
            : Path.of("..", "query.sql");
    return Files.readString(path).trim();
  }

  private static String envOrNull(String name) {
    String value = System.getenv(name);
    if (value == null || value.isBlank()) {
      return null;
    }
    return value.trim();
  }

  private static String envOrDefault(String name, String defaultValue) {
    String value = envOrNull(name);
    return value != null ? value : defaultValue;
  }

  private static void printUsage() {
    System.err.println(
        "usage: AnalyzeAndRender [options]\n"
            + "  --query-mode PLAN|PROFILE   Spanner execute-sql mode (default: PLAN)\n"
            + "  --project PROJECT           GCP project id\n"
            + "  --instance INSTANCE         Spanner instance id\n"
            + "  --database DATABASE         Spanner database id\n"
            + "  --query SQL                 SQL text (overrides --query-file)\n"
            + "  --query-file PATH           SQL file (default: ../query.sql)\n"
            + "\n"
            + "Environment (when flags omitted):\n"
            + "  SPANNER_QUERY_MODE, SPANNER_PROJECT_ID, SPANNER_INSTANCE_ID,\n"
            + "  SPANNER_DATABASE_ID, SPANNER_QUERY, SPANNER_QUERY_FILE");
  }
}
